### PR TITLE
tools/ceph-objectstore-tool: make dump-export optionally dump object data

### PR DESCRIPTION
**Pull request by trociny**
If --dump-data-dir option is specified object data is dumped to
the object file in that directory.

This may be useful when one wants to recover data from a dead
cluster.

Signed-off-by: Mykola Golub <mykola.golub@clyso.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by trociny** [comment]
jenkins test make check

File: src/tools/ceph_objectstore_tool.cc [hunk 1]
```diff
@@ -1069,7 +1070,23 @@ int dump_data(Formatter *formatter, bufferlist &bl)
...
   formatter->dump_unsigned("offset", ds.offset);
   formatter->dump_unsigned("len", ds.len);
-  // XXX: Add option to dump data like od -cx ?
+  if (!dump_data_path.empty()) {
+    int fd = open(dump_data_path.c_str(), O_WRONLY|O_CREAT|O_LARGEFILE, 0666);
+    if (fd == -1) {
+      std::cerr << "open " << dump_data_path << " failed: "
+                << cpp_strerror(errno) << std::endl;
+    } else {
+      int ret = pwrite(fd, ds.databl.c_str(), ds.len, ds.offset);
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:1079]: Hmm, I'm not 100% sure but I think I can recall some methods in `bufferlist` for dumping its content to a file. Perhaps we could reuse?

**Review by rzarzynski — APPROVED** [review]
Basically LGTM; just one minor thing.

File: src/tools/ceph_objectstore_tool.cc (hunk 1 above)
Comment (trociny) [reply src/tools/ceph_objectstore_tool.cc]: @rzarzynski Thanks. Updated

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -1146,7 +1147,23 @@ int dump_data(Formatter *formatter, bufferlist &bl)
...
   formatter->dump_unsigned("offset", ds.offset);
   formatter->dump_unsigned("len", ds.len);
-  // XXX: Add option to dump data like od -cx ?
+  if (!dump_data_path.empty()) {
+    int fd = open(dump_data_path.c_str(), O_WRONLY|O_CREAT|O_LARGEFILE, 0666);
+    if (fd == -1) {
+      std::cerr << "open " << dump_data_path << " failed: "
+                << cpp_strerror(errno) << std::endl;
+    } else {
+      int ret = ds.databl.write_fd(fd, ds.offset);
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:1156]: :+1:

**Review by rzarzynski — APPROVED** [review]

**Comment by ljflores** [comment]
Rados suite review: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomclM1xjBe01744-wip-yuri-testing-2023-05-10-0917

**Pull request merged**